modernize(tags): list tags api [#32960]

hassandotcms opened the pull request:
Proposed Changes:
- GET /api/v2/tags returns all tags (up to default limit)
- filter=market returns tags containing "market" (case-insensitive)
- global=true includes SYSTEM_HOST tags, false excludes them
- site parameter filters by site ID or name correctly
- Can combine filter + site + global parameters

This PR fixes: #32960

This PR fixes: #32960

hassandotcms committed:
refactor(tags): modernize tag API v2 endpoints for REST compliance

- Move single tag creation from POST /api/v2/tags/single to POST /api/v2/tags
- Move bulk tag creation from POST /api/v2/tags to POST /api/v2/tags/_bulk
- Add SingleTagForm with Bean Validation for single tag requests
- Add ResponseEntityRestTagView for type-safe single tag responses
- Implement SYSTEM_HOST fallback for cross-site tag retrieval
- Add comprehensive OpenAPI annotations for all endpoints
- Add custom validation for business rules (comma check, whitespace)
- Update Postman tests for new endpoint paths and response formats
- Fix test expectations to match actual API behavior (tagStorage, validation)

This refactoring aligns the Tags API with REST best practices where the base
resource path handles single entity operations and special suffixes like /_bulk
handle batch operations. Includes proper error handling, validation, and
cross-site tag discovery through SYSTEM_HOST fallback.

Refs: #32966

hassandotcms committed:
feat(tags): modernize PUT/DELETE endpoints for v2 Tags API

- Add comprehensive PUT /api/v2/tags/{idOrName} endpoint with dual identification support
- Add proper DELETE /api/v2/tags/{tagId} endpoint returning 204 No Content
- Update UpdateTagForm with @JsonAlias to support modern field names (name/tagName)
- Add strict site validation without fallback for PUT operations
- Include comprehensive OpenAPI documentation with proper error responses
- Update Postman tests for new endpoint structure and validation

hassandotcms committed:
feat(tags): modernize v2 list API with pagination and filtering

- Implement modern pagination with ResponseEntityPaginatedDataView supporting page/per_page parameters
- Add case-insensitive filtering with LENGTH() ordering for shortest matches first
- Support flexible site parameter resolution (UUID, name, or "SYSTEM_HOST" literal)
- Add global parameter to control SYSTEM_HOST tag inclusion
- Update OpenAPI documentation with comprehensive parameter descriptions and response schemas
- Fix tag storage handling bug where SYSTEM_HOST sites incorrectly returned global tags when global=false
- Add getEffectiveTagStorage() helper method to centralize tag storage resolution logic
- Update Postman tests with modern parameters and fix case insensitive test with global=true

Breaking changes: Response format changed from Map<String, RestTag> to paginated array structure
with entity, pagination metadata, and proper REST compliance.

hassandotcms committed:
modernize(tags): modernize create/get tags api

Make create tag idempotent, will return tag if already existing (201)

Refs: #32966

hassandotcms committed:
modernize(tags): modernize create/get tags api

move validation to a single place. be consistent where we do validations.
ref: #32966

hassandotcms committed:
modernize(tags): modernize create/get tags api

 use ResponseEntityView for proper and consistent response of errors and exceptions.
 ref: #32966

hassandotcms committed:
modernize(tags): modernize create/get tags api

 cleanup. remove extra comments.
 ref: #32966

hassandotcms committed:
modernize(tags): modernize create/get tags api

fix postman tests, and verify assertions.
ref: #32966

hassandotcms committed:
Merge branch 'main' of https://github.com/dotCMS/core into 32966-modernize-post-create-tag-apis

hassandotcms committed:
modernize(tags): modernize create/get tags api

 update category resource postman test to use correct tags api.
 ref: #32966

hassandotcms committed:
modernize(tags): 32966-modernize-create-get-tag-apis

removed explicit exception catches. add debug statements.
Ref: #32966

hassandotcms committed:
Merge branch 'main' of https://github.com/dotCMS/core into 32966-modernize-post-create-tag-apis

hassandotcms committed:
Merge branch '32966-modernize-post-create-tag-apis' into 32941-implement-put-delete-tag-endpoints

hassandotcms committed:
modernize(tags): put/delete tag apis

revert to original delete api.
ref: #32967

hassandotcms committed:
modernize(tags): modernize DELETE and UPDATE APIs

  - Revert DELETE to throw DoesNotExistException instead of manual Response building
  - Refactor UPDATE to use form validation and exception throwing
  - Update Postman tests to match new response formats
 ref: #32967

hassandotcms committed:
modernize(tags): put/delete tag apis

 added comments.
 ref: #32967

hassandotcms committed:
Merge branch '32966-modernize-post-create-tag-apis' into 32960-modernize-list-tags-api

hassandotcms committed:
Merge branch '32967-modernize-put-delete-tag-endpoints' into 32960-modernize-list-tags-api

hassandotcms committed:
Merge branch 'main' of https://github.com/dotCMS/core into 32960-modernize-list-tags-api

hassandotcms committed:
Merge branch 'main' into 32960-modernize-list-tags-api

hassandotcms committed:
modernize(tags): modernize list tag apis

remove explicit exception handling.
ref: #32960

hassandotcms committed:
Merge branch '32960-modernize-list-tags-api' of https://github.com/dotCMS/core into 32960-modernize-list-tags-api

hassandotcms committed:
modernize(tags): put/delete tag apis

return view instead of Response.
ref: #32967

hassandotcms committed:
modernize(tags): modernize create/get tags apis

fix typo.
ref: #32966

hassandotcms committed:
modernize(tags): unify single and bulk tag creation into list-based API

  - Replace separate endpoints with unified POST /api/v2/tags
  - Fix validation errors to return consistent ErrorEntity array format
  - Add proper Swagger documentation with ResponseEntityRestTagListView
  - Update all Postman tests to match new API behavior
ref: #32967

hassandotcms committed:
Merge branch '32966-create-tags-multiple' into 32967-modernize-put-delete-tag-endpoints

hassandotcms committed:
modernize(tags): unify single and bulk tag creation into list-based API

 updated api docs.
 ref: #32967

hassandotcms committed:
modernize(tags): unify single and bulk tag creation into list-based API

remove wildcard imports.
ref: #32967

hassandotcms committed:
modernize(tags): unify single and bulk tag creation into list-based API

 use responseviewlisttagsview in response entity.
 ref: #32967

hassandotcms committed:
modernize(tags): unify single and bulk tag creation into list-based API

fix postman failing test in content resource.
ref: #32967

hassandotcms committed:
Merge branch 'main' of https://github.com/dotCMS/core into 32967-modernize-put-delete-tag-endpoints

hassandotcms committed:
modernize(tags): tags api

  - Create TagValidationHelper for DRY validation logic
  - Move SingleTagForm to v2 package with shared validation
  - Add v2-specific UpdateTagForm with consistent validation
  - Use StringPool.COMMA for consistency
ref: #32967

hassandotcms committed:
modernize(tags): tags api

refactoring. remove singletagform.
ref: #3i2967

hassandotcms committed:
modernize(tags): tags api

keep v1 updatesingleform intact
ref: #3i2967

hassandotcms committed:
Merge branch 'main' into 32967-modernize-put-delete-tag-endpoints

hassandotcms committed:
Merge branch '32967-modernize-put-delete-tag-endpoints' into 32960-modernize-list-tags-api

hassandotcms committed:
modernize(tags): list tags api

fix breaking postman tests.
ref: #32960

hassandotcms committed:
modernize(tags): list tags api with pagination

  - Add comprehensive Postman tests for site-specific tag storage
  - Verify tag isolation with global=true/false parameter behavior
  - Test LIKE escaping for special characters in tag searches
ref: #32960

hassandotcms committed:
Merge branch 'main' of https://github.com/dotCMS/core into 32960-modernize-list-tags-api

hassandotcms committed:
modernize(tags): list tags api with pagination

  - make list tags tests reliable and independent of other tests data.
ref: #32960

hassandotcms committed:
Merge branch 'main' of https://github.com/dotCMS/core into 32960-modernize-list-tags-api

hassandotcms committed:
modernize(tags): list tags api with pagination
  - refactor: remove magic number, improve character escaping in query, fix imports.
ref: #32960

hassandotcms committed:
Merge branch 'main' of https://github.com/dotCMS/core into 32960-modernize-list-tags-api

hassandotcms committed:
modernize(tags): list tags api with pagination
  - refactor: fix direction query param
ref: #32960

jdotcms commented on dotCMS/src/main/java/com/dotcms/rest/api/v2/tags/TagResource.java:
```diff
@@ -97,43 +105,155 @@ protected TagResource(final TagAPI tagAPI, final HostAPI hostAPI, final FolderAP
-                ? String.format("Listing Tag(s) '%s' from Site '%s'", tagName, siteId)
-                : "Listing ALL Tags");
+        // 2. Validate parameters
+        validatePaginationParams(page, perPage);
```
We already have a pattern to do the pagination, see PaginationUtil and all the Paginator that use it

hassandotcms replied:
resolved.

jdotcms commented on dotCMS/src/main/java/com/dotmarketing/tag/business/TagFactoryImpl.java:
```diff
@@ -328,6 +320,100 @@ public List<Tag> getFilteredTags(String tagName, String hostFilter, boolean glob
     }
 
+    @Override
+    public long getFilteredTagsCount(String tagName, String hostFilter, boolean globalTagsFilter, boolean excludePersonas) throws DotDataException {
```
may boolean flags is an antipattern it would be better to use a bean + builder

hassandotcms committed:
modernize(tags): use PaginationUtil pattern for v2 list endpoint

  - Implement TagsPaginator with standard pagination pattern
  - Add debug logging to getFilteredTagsCount()
  - Delegate pagination handling to PaginationUtil
ref: #32960

hassandotcms committed:
modernize(tags): list tags api with pagination
   - fix postman tests to comply with paginationUtil
 ref: #32960

hassandotcms committed:
Merge branch 'main' of https://github.com/dotCMS/core into 32960-modernize-list-tags-api

hassandotcms commented on dotCMS/src/main/java/com/dotmarketing/tag/business/TagAPIImpl.java:
```diff
@@ -66,6 +66,12 @@ public List<Tag> getFilteredTags ( String tagName, String hostFilter, boolean gl
+    @CloseDBIfOpened
+    @Override
+    public long getFilteredTagsCount(String tagName, String hostFilter, boolean globalTagsFilter) throws DotDataException {
+        return tagFactory.getFilteredTagsCount(tagName, hostFilter, globalTagsFilter, true);
```
added

hassandotcms committed:
Merge branch 'main' into 32960-modernize-list-tags-api
